// @shopify/jest-koa-mocks/build/cjs/create-mock-context/create-mock-context.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');
var url = require('url');
var stream = require('stream');
var httpMocks = require('node-mocks-http');
var Koa = require('koa');
var createMockCookies = require('../create-mock-cookies/create-mock-cookies.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var stream__default = /*#__PURE__*/_interopDefaultLegacy(stream);
var httpMocks__default = /*#__PURE__*/_interopDefaultLegacy(httpMocks);
var Koa__default = /*#__PURE__*/_interopDefaultLegacy(Koa);

function createContext(options = {}) {
  const app = new Koa__default["default"]();
  const {
    cookies,
    method,
    statusCode,
    session,
    requestBody,
    rawBody = '',
    url: url$1 = '',
    host = 'test.com',
    encrypted = false,
    throw: throwFn = jest.fn(),
    redirect = jest.fn(),
    headers = {},
    state = {},
    customProperties = {}
  } = options;

  const extensions = _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, customProperties), {}, {
    throw: throwFn,
    session,
    redirect,
    state
  });

  const protocolFallback = encrypted ? 'https' : 'http';
  const urlObject = new url.URL(url$1, `${protocolFallback}://${host}`);
  const req = httpMocks__default["default"].createRequest({
    url: urlObject.toString(),
    method,
    statusCode,
    session,
    headers: _rollupPluginBabelHelpers.objectSpread2({
      // Koa determines protocol based on the `Host` header.
      Host: urlObject.host
    }, headers)
  }); // Some functions we call in the implementations will perform checks for `req.encrypted`, which delegates to the socket.
  // MockRequest doesn't set a fake socket itself, so we create one here.

  req.socket = new stream__default["default"].Duplex();
  Object.defineProperty(req.socket, 'encrypted', {
    writable: false,
    value: urlObject.protocol === 'https:'
  });
  const res = httpMocks__default["default"].createResponse(); // Koa sets a default status code of 404, not the node default of 200
  // https://github.com/koajs/koa/blob/master/docs/api/response.md#responsestatus

  res.statusCode = 404; // This is to get around an odd behavior in the `cookies` library, where if `res.set` is defined, it will use an internal
  // node function to set headers, which results in them being set in the wrong place.

  res.set = undefined;
  const context = app.createContext(req, res);
  Object.assign(context, extensions);
  context.cookies = createMockCookies["default"](cookies); // ctx.request.body is a common enough custom property for middleware to add that it's handy to just support it by default

  context.request.body = requestBody;
  context.request.rawBody = rawBody;
  return context;
}

exports["default"] = createContext;
