// flatpickr/dist/esm/l10n/fi.js
const fp = typeof window !== "undefined" && window.flatpickr !== undefined
    ? window.flatpickr
    : {
        l10ns: {},
    };
export const Finnish = {
    firstDayOfWeek: 1,
    weekdays: {
        shorthand: ["Su", "Ma", "Ti", "Ke", "To", "Pe", "La"],
        longhand: [
            "Sunnuntai",
            "Maanantai",
            "Tiistai",
            "Keskiviikko",
            "Torstai",
            "Perjantai",
            "Lauantai",
        ],
    },
    months: {
        shorthand: [
            "Tammi",
            "Helmi",
            "Maalis",
            "Huhti",
            "Touko",
            "Kes채",
            "Hein채",
            "Elo",
            "Syys",
            "Loka",
            "Marras",
            "Joulu",
        ],
        longhand: [
            "Tammikuu",
            "Helmikuu",
            "Maaliskuu",
            "Huhtikuu",
            "Toukokuu",
            "Kes채kuu",
            "Hein채kuu",
            "Elokuu",
            "Syyskuu",
            "Lokakuu",
            "Marraskuu",
            "Joulukuu",
        ],
    },
    ordinal: () => {
        return ".";
    },
    time_24hr: true,
};
fp.l10ns.fi = Finnish;
export default fp.l10ns;
